knora/dsplib/utils/xml_upload.py

```diff
@@ -550,20 +594,57 @@ def do_sort_order(resources: list[XMLResource], verbose) -> list[XMLResource]:
-            exit(1)
+            # there are circular references. go through all problematic resources, and stash the problematic references.
+            # there are two types, text and resptr, which need to be treated differently.
+            for res in nok_resources.copy():
+                for link_prop in res.get_props_with_links():
+                    if link_prop.valtype == 'text':
+                        for value in link_prop.values:
+                            if value.resrefs and not all([_id in ok_res_ids for _id in value.resrefs]):
+                                # stash this XML text, replace it by its hash, and remove the
+                                # problematic resrefs from the XMLValue's resrefs list
+                                value_hash = str(hash(f'{value.value}{datetime.now()}'))
+                                if res not in stashed_xml_texts:
+                                    stashed_xml_texts[res] = {link_prop: {value_hash: cast(KnoraStandoffXml, value.value)}}
+                                elif link_prop not in stashed_xml_texts[res]:
+                                    stashed_xml_texts[res][link_prop] = {value_hash: cast(KnoraStandoffXml, value.value)}
+                                else:
+                                    stashed_xml_texts[res][link_prop][value_hash] = cast(KnoraStandoffXml, value.value)
+                                value.value = KnoraStandoffXml(value_hash)
+                                value.resrefs = [_id for _id in value.resrefs if _id in ok_res_ids]
+                    elif link_prop.valtype == 'resptr':
+                        for value in link_prop.values.copy():
+                            if value.value not in ok_res_ids:
+                                # value.value is the id of the target resource. stash it, then delete it
+                                if res not in stashed_resptr_props:
+                                    stashed_resptr_props[res] = {}
+                                    stashed_resptr_props[res][link_prop] = [str(value.value)]
+                                else:
+                                    if link_prop not in stashed_resptr_props[res]:
+                                        stashed_resptr_props[res][link_prop] = [str(value.value)]
+                                    else:
+                                        stashed_resptr_props[res][link_prop].append(str(value.value))
+                                link_prop.values.remove(value)
+                    else:
+                        # This should never happen. If this code is ever executed, it proves that there is a
+                        # misconception in the underlying assumptions of the code logic
+                        print(f'Warning in remove_circular_references(): link_prop.valtype is neither text '
+                              f'nor resptr. Mistake in the underlying assumptions of the code logic.')
```

Comment:
This seems weird to me. If this should never happen, the code should be removed or at least an error returned / exception thrown.

Reply:
ok, I agree. I raised a BaseError instead